### PR TITLE
MGMT-9341: Interactive text checker in the cloud

### PR DESCRIPTION
**Pull request by jgyselov**
https://issues.redhat.com/browse/MGMT-9341

image
image
image
image

### CODE REVIEW

File: src/common/components/ui/formik/validationSchemas.ts [hunk 1]
```diff
@@ -9,7 +9,7 @@ import { ProxyFieldsType } from '../../../types';
 import { trimCommaSeparatedList, trimSshPublicKey } from './utils';
 
 const ALPHANUMBERIC_REGEX = /^[a-zA-Z0-9]+$/;
-const CLUSTER_NAME_REGEX = /^[a-z]([-a-z0-9]*[a-z0-9])?$/;
+const CLUSTER_NAME_REGEX = /^[a-z](.*[a-z0-9])?$/;
```
Comment (jkilzi) [inline src/common/components/ui/formik/validationSchemas.ts:12]: `.*` means that the name can contain special characters in the middle (e.g. $%^&*#...)
Why is this change necessary?

File: src/common/components/ui/formik/validationSchemas.ts [hunk 2]
```diff
@@ -22,28 +22,47 @@ const MAC_REGEX = /^([0-9A-Fa-f]{2}[:-]){5}([0-9A-Fa-f]{2})|([0-9a-fA-F]{4}\\.[0
 const BMC_REGEX = /^((ipmi|ibmc(\+https?)?|idrac(\+https?)?|idrac-redfish(\+https?)?|idrac-virtualmedia(\+https?)?|irmc|redfish(\+https?)?|redfish-virtualmedia(\+https?)?|ilo4(\+https)?|ilo4-virtuallmedia(\+https)?|ilo5(\+https)?|ilo5-redfish(\+https)|ilo5-virtualmedia(\+https)|https?|ftp):(\/\/([a-z0-9\-._~%!$&'()*+,;=]+@)?([a-z0-9\-._~%]+|\[[a-f0-9:.]+\]|\[v[a-f0-9][a-z0-9\-._~%!$&'()*+,;=:]+\])(:[0-9]+)?(\/[a-z0-9\-._~%!$&'()*+,;=:@]+)*\/?|(\/?[a-z0-9\-._~%!$&'()*+,;=:@]+(\/[a-z0-9\-._~%!$&'()*+,;=:@]+)*\/?)?)|([a-z0-9\-._~%!$&'()*+,;=@]+(\/[a-z0-9\-._~%!$&'()*+,;=:@]+)*\/?|(\/[a-z0-9\-._~%!$&'()*+,;=:@]+)+\/?))(\?[a-z0-9\-._~%!$&'()*+,;=:@/?]*)?(#[a-z0-9\-._~%!$&'()*+,;=:@/?]*)?$/i;
 const LOCATION_CHARS_REGEX = /^[a-zA-Z0-9-._]*$/;
 
+export const CLUSTER_NAME_VALIDATION_MESSAGES = {
+  INVALID_LENGTH: '1-54 characters',
+  INVALID_VALUE: 'Use alphanumberic characters, or hyphen (-)',
+  INVALID_START_END: 'Must start with a letter and end with a letter or number.',
+};
+
+export const UNIQUE_CLUSTER_NAME_VALIDATION_MESSAGES = {
```
Comment (jkilzi) [inline src/common/components/ui/formik/validationSchemas.ts:31]:
```suggestion
export const uniqueClusterNameValidationMessages = {
```

File: src/common/components/ui/formik/validationSchemas.ts [hunk 3]
```diff
@@ -22,28 +22,47 @@ const MAC_REGEX = /^([0-9A-Fa-f]{2}[:-]){5}([0-9A-Fa-f]{2})|([0-9a-fA-F]{4}\\.[0
 const BMC_REGEX = /^((ipmi|ibmc(\+https?)?|idrac(\+https?)?|idrac-redfish(\+https?)?|idrac-virtualmedia(\+https?)?|irmc|redfish(\+https?)?|redfish-virtualmedia(\+https?)?|ilo4(\+https)?|ilo4-virtuallmedia(\+https)?|ilo5(\+https)?|ilo5-redfish(\+https)|ilo5-virtualmedia(\+https)|https?|ftp):(\/\/([a-z0-9\-._~%!$&'()*+,;=]+@)?([a-z0-9\-._~%]+|\[[a-f0-9:.]+\]|\[v[a-f0-9][a-z0-9\-._~%!$&'()*+,;=:]+\])(:[0-9]+)?(\/[a-z0-9\-._~%!$&'()*+,;=:@]+)*\/?|(\/?[a-z0-9\-._~%!$&'()*+,;=:@]+(\/[a-z0-9\-._~%!$&'()*+,;=:@]+)*\/?)?)|([a-z0-9\-._~%!$&'()*+,;=@]+(\/[a-z0-9\-._~%!$&'()*+,;=:@]+)*\/?|(\/[a-z0-9\-._~%!$&'()*+,;=:@]+)+\/?))(\?[a-z0-9\-._~%!$&'()*+,;=:@/?]*)?(#[a-z0-9\-._~%!$&'()*+,;=:@/?]*)?$/i;
 const LOCATION_CHARS_REGEX = /^[a-zA-Z0-9-._]*$/;
 
+export const CLUSTER_NAME_VALIDATION_MESSAGES = {
```
Comment (jkilzi) [inline src/common/components/ui/formik/validationSchemas.ts:25]: Since this is just an object I think it would be more appropriate to use camelCase

```suggestion
export const clusterNameValidationMessages = {
```

File: src/common/components/ui/formik/validationSchemas.ts (hunk 2 above)
Comment (jkilzi) [reply src/common/components/ui/formik/validationSchemas.ts:31]: I would even suggest moving these messages to a different file...

**Review by jkilzi — CHANGES_REQUESTED** [review]

File: src/common/components/ui/formik/validationSchemas.ts (hunk 1 above)
Comment (jgyselov) [reply src/common/components/ui/formik/validationSchemas.ts:12]: I wanted the conditions in the list to be simple, straight forward and always deal with one thing at a time.
This regex was good but it tested two things at a time and that didn't allow for separation of the conditions ("you can use these characters" and "name must start and end with ..."). I think that for the users a slightly longer list with shorter instructions is easier to comprehend than a shorter list with long and complicated instructions.

File: src/common/components/ui/formik/validationSchemas.ts (hunk 2 above)
Comment (jgyselov) [reply src/common/components/ui/formik/validationSchemas.ts:32]: I was following a "convetion" we had with the other validation messages...
But I do agree, moving all of them to a different file might be a good idea.

**Comment by nirfarkas** [comment]
Looks great!

**Review by jkilzi — COMMENTED** [review]
LGTM
@rawagner PTAL

**Comment by jkilzi** [comment]
@jgyselov we need to cover this with integration tests

**Comment by jkilzi** [comment]
@jgyselov please check if the cluster name can contain capital letters, your regex won't allow it (also the previous one) but the rules don't explicitly say the characters must be all lowercase letters

**Comment by jgyselov** [comment]
> @jgyselov please check if the cluster name can contain capital letters, your regex won't allow it (also the previous one) but the rules don't explicitly say the characters must be all lowercase letters

BE response: 
```
Cluster name format is not valid: 'CLUSTER'. Name must consist of lower-case letters, numbers and hyphens.
It must start with a letter and end with a letter or number.
```
I've modified the regexes and validation messages to reflect that only lowercase letters are allowed for both cluster names and host names.
@jkilzi

**Comment by jkilzi** [comment]
@jgyselov can you ptal, there are merge conflicts

**Comment by jgyselov** [comment]
> can you ptal, there are merge conflicts

Resolved. @jkilzi

**Review by jkilzi — APPROVED** [review]

**Comment by jkilzi** [comment]
@jgyselov we need to port this to releases/v2.4
